datafusion/proto/src/logical_plan.rs

```diff
@@ -760,55 +804,21 @@ impl AsLogicalPlan for LogicalPlanNode {
-                    let file_format_type = if let Some(parquet) =
-                        any.downcast_ref::<ParquetFormat>()
-                    {
-                        FileFormatType::Parquet(protobuf::ParquetFormat {
-                            enable_pruning: parquet.enable_pruning(),
-                        })
-                    } else if let Some(csv) = any.downcast_ref::<CsvFormat>() {
-                        FileFormatType::Csv(protobuf::CsvFormat {
-                            delimiter: byte_to_string(csv.delimiter())?,
-                            has_header: csv.has_header(),
-                        })
-                    } else if any.is::<AvroFormat>() {
-                        FileFormatType::Avro(protobuf::AvroFormat {})
-                    } else {
-                        return Err(proto_error(format!(
-                            "Error converting file format, {:?} is invalid as a datafusion foramt.",
-                            listing_table.options().format
-                        )));
-                    };
-                    Ok(protobuf::LogicalPlanNode {
-                        logical_plan_type: Some(LogicalPlanType::ListingScan(
-                            protobuf::ListingTableScanNode {
-                                file_format_type: Some(file_format_type),
-                                table_name: table_name.to_owned(),
-                                collect_stat: listing_table.options().collect_stat,
-                                file_extension: listing_table
-                                    .options()
-                                    .file_extension
-                                    .clone(),
-                                table_partition_cols: listing_table
-                                    .options()
-                                    .table_partition_cols
-                                    .clone(),
-                                paths: listing_table
-                                    .table_paths()
-                                    .iter()
-                                    .map(|x| x.to_string())
-                                    .collect(),
-                                schema: Some(schema),
-                                projection,
-                                filters,
-                                target_partitions: listing_table
-                                    .options()
-                                    .target_partitions
-                                    as u32,
-                            },
-                        )),
-                    })
+                    Self::serialize_listing_table(
+                        table_name,
+                        projection,
+                        schema,
+                        filters,
+                        listing_table,
+                    )
+                } else if let Some(custom_table) = source.downcast_ref::<CustomTable>() {
```

Comment:
So this seems to be the key issue -- if `source` isn't some type that DataFusion knows about it can't do a useful job serializing it.

I don't fully understand how a `CustomTable` struct will help -- at some point the client code needs to get the concrete  `TableProvider`, not a `CustomTable`.

I think there was a similar issue serializing "extensions" for which @thinkharderdev  added 

```rust
pub trait LogicalExtensionCodec: Debug + Send + Sync {
```

I wonder if we can follow the same model here , something like add `LogicalExtensionCodec::try_encoded_table_provider` and `LogicalExtensionCodec::try_decode_table_provider`


```rust
if let Some(listing_table) = source.downcast_ref::<ListingTable>() {
  // ...
} else {
  // delegate to provided extension codec:
  extension_codec.try_encode_table_provider(
                table_name,
                source,
                filters,
                projection,
  ); 
} 
```

Reply:
> I don't fully understand this proposal 

That delta PR really explains what I'm trying to do. Unfortunately the end goal spans datafusion, delta-rs & ballista, so it's hard to convey. And I'm open to doing it simpler ways if they exist and fulfill the broader goal of supporting table types unknown to datafusion (i.e. delta, but possibly others).

> the client code needs to get the concrete TableProvider

As long as the deserializing node also has the `TableProviderFactory` registered, it can get a concrete one created. The CustomTable is just a wrapper that contains enough information to get it instantiated by the factory.

Reply:
> I wonder if we can follow the same model here

Reviewing...

Reply:
```
if let Some(listing_table) = source.downcast_ref::<ListingTable>() {
  // ...
} else {
```

I don't think this would fit with the goal, as datafusion would have to include types at compile time for each table, i.e. we'd have to make a compile time dependency on delta-rs which itself has a compile-time dependency on datafusion already.

Reply:
> have to include types at compile time

Unless that `downcast_ref::` trick works on traits as well as structs.

Reply:
I see it works via the `dyn Any` trait, so that seems okay. I was worried about having to add `delta-rs` stuff to the `datafusion.proto`, but I see `pub struct TopKPlanProto` gets away without modifying the protobuf definition. If you think this will work @alamb I'll start work on another PR doing it with the codec instead.

Reply:
> will have to provide its own serialization / deserialization

That works for me!

Reply:
I think we'll need something like: 

```
message LogicalPlanNode {
  oneof LogicalPlanType {
    CustomTableScanNode custom_scan = 25;
```
and
```
message CustomTableScanNode {
  bytes custom_table_data = 1;
}
```

Reply:
Actually, it looks like we are conflating the `scan` data with the `TableProvider` data when we serialize. I split them out in `is a stub for testing`.